File: ios_dashboard/SandboxedDashboard/Models/Ask.swift
import Foundation

// Models for the Ask assistant — the non-interrupting sidecar co-pilot.
// These mirror the backend `ask_threads` / `ask_messages` payloads exactly.
// (The decoder uses plain JSONDecoder with no key strategy, so every snake_case
// field needs an explicit CodingKey.)

// MARK: - Thread

struct AskThread: Codable, Identifiable {
    let id: String
    let missionId: String
    let title: String?
    let model: String?
    let createdAt: String
    let updatedAt: String

    enum CodingKeys: String, CodingKey {
        case id, title, model
        case missionId = "mission_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    var displayTitle: String { title ?? "Untitled thread" }
}

// MARK: - Message

struct AskMessage: Codable, Identifiable {
    let id: String
    let threadId: String
    let seq: Int
    /// "user" | "assistant" | "tool_call" | "tool_result"
    let role: String
    let content: String
    let toolName: String?
    let toolCallId: String?
    let createdAt: String
    /// Client-only delivery state for optimistic user bubbles. Excluded from
    /// `CodingKeys`, so it's never decoded from / encoded to the backend and
    /// defaults to `.sent` for server-sourced messages. Mirrors the main
    /// composer's `ChatMessage.sendState` so Ask gets the same
    /// pending/failed/tap-to-retry UX.
    var sendState: MessageSendState = .sent

    enum CodingKeys: String, CodingKey {
        case id, seq, role, content
        case threadId = "thread_id"
        case toolName = "tool_name"
        case toolCallId = "tool_call_id"
        case createdAt = "created_at"
        // `metadata` and `sendState` are intentionally omitted — arbitrary
        // JSON we don't render, and a client-only field, respectively.
    }

    var isUser: Bool { role == "user" }
    var isAssistant: Bool { role == "assistant" }
    var isToolCall: Bool { role == "tool_call" }
    var isToolResult: Bool { role == "tool_result" }
    var isTool: Bool { isToolCall || isToolResult }
}

// MARK: - API responses

/// Response of `POST /api/control/missions/:id/ask`.
struct AskSendResponse: Codable {
    let threadId: String
    let answer: String
    let messages: [AskMessage]

    enum CodingKeys: String, CodingKey {
        case answer, messages
        case threadId = "thread_id"
    }
}

/// Response of `GET /api/control/missions/:id/ask/threads/:tid`
/// (thread fields flattened alongside `messages`).
struct AskThreadDetail: Codable {
    let id: String
    let missionId: String
    let title: String?
    let model: String?
    let createdAt: String
    let updatedAt: String
    let messages: [AskMessage]

    enum CodingKeys: String, CodingKey {
        case id, title, model, messages
        case missionId = "mission_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

/// One Server-Sent Event from `POST /api/control/missions/:id/ask/stream`.
/// A single shape covers every variant (`type` discriminates).
struct AskStreamEvent: Decodable {
    let type: String
    let content: String?
    let toolCallId: String?
    let name: String?
    let args: String?
    let result: String?
    let threadId: String?
    let answer: String?
    let message: String?

    enum CodingKeys: String, CodingKey {
        case type, content, name, args, result, answer, message
        case toolCallId = "tool_call_id"
        case threadId = "thread_id"
    }
}

/// Error carrying a co-pilot stream failure message (from an SSE `error` event
/// or an abruptly-ended stream), so it surfaces through the normal throw path.
struct AskStreamError: LocalizedError {
    let message: String
    var errorDescription: String? { message }
}

File: android_dashboard/app/src/main/java/sh/sandboxed/dashboard/data/Ask.kt
package sh.sandboxed.dashboard.data

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient

// Models for the Ask assistant — the non-interrupting sidecar co-pilot.
// These mirror the backend `ask_threads` / `ask_messages` payloads (and the
// iOS Ask.swift wire contract) exactly. Net.json uses snake_case field names,
// so every multi-word field needs an explicit @SerialName.

// ---- Client-only delivery state for optimistic user bubbles ----

/// Mirrors the main composer's send state so Ask gets the same
/// pending / failed / tap-to-retry UX. Never serialized.
sealed class AskSendState {
    data object Sent : AskSendState()
    data object Pending : AskSendState()
    data class Failed(val reason: String) : AskSendState()

    val isPending: Boolean get() = this is Pending
    val isFailed: Boolean get() = this is Failed
}

// ---- Thread ----

@Serializable
data class AskThread(
    val id: String,
    @SerialName("mission_id") val missionId: String = "",
    val title: String? = null,
    val model: String? = null,
    @SerialName("created_at") val createdAt: String = "",
    @SerialName("updated_at") val updatedAt: String = "",
) {
    val displayTitle: String get() = title?.takeIf { it.isNotBlank() } ?: "Untitled thread"
}

// ---- Message ----

@Serializable
data class AskMessage(
    val id: String,
    @SerialName("thread_id") val threadId: String = "",
    val seq: Int = 0,
    /// "user" | "assistant" | "tool_call" | "tool_result"
    val role: String,
    val content: String = "",
    @SerialName("tool_name") val toolName: String? = null,
    @SerialName("tool_call_id") val toolCallId: String? = null,
    @SerialName("created_at") val createdAt: String = "",
    // Client-only — defaults to Sent for server-sourced messages.
    @Transient val sendState: AskSendState = AskSendState.Sent,
) {
    val isUser: Boolean get() = role == "user"
    val isAssistant: Boolean get() = role == "assistant"
    val isToolCall: Boolean get() = role == "tool_call"
    val isToolResult: Boolean get() = role == "tool_result"
    val isTool: Boolean get() = isToolCall || isToolResult
}

// ---- API payloads ----

/// Request body for `POST /api/control/missions/:id/ask[/stream]`.
@Serializable
data class AskStreamRequest(
    val content: String,
    @SerialName("thread_id") val threadId: String? = null,
)

/// Response of `GET /api/control/missions/:id/ask/threads/:tid`
/// (thread fields flattened alongside `messages`).
@Serializable
data class AskThreadDetail(
    val id: String,
    @SerialName("mission_id") val missionId: String = "",
    val title: String? = null,
    val model: String? = null,
    @SerialName("created_at") val createdAt: String = "",
    @SerialName("updated_at") val updatedAt: String = "",
    val messages: List<AskMessage> = emptyList(),
)

/// One Server-Sent Event from `POST /api/control/missions/:id/ask/stream`.
/// A single shape covers every variant (`type` discriminates).
@Serializable
data class AskStreamEvent(
    val type: String,
    val content: String? = null,
    @SerialName("tool_call_id") val toolCallId: String? = null,
    val name: String? = null,
    val args: String? = null,
    val result: String? = null,
    @SerialName("thread_id") val threadId: String? = null,
    val answer: String? = null,
    val message: String? = null,
)

/// Carries a co-pilot stream failure (an SSE `error` event or an abruptly
/// ended stream) so it surfaces through the normal Flow error path.
class AskStreamException(message: String) : RuntimeException(message)
